[7-1-stable] Fix a crash when an  Executor wrapped fork exit.

JoeDupuis opened the pull request:
Fix #51298

forking inside an Execution Wrapper crashes when running the completion callbacks.

Rails 7.0 was wrapping the `runner` command with an ExecutionWrapper. 
Rails 7.2 changed the definition of `active_connection?`, the new definition doesn't contain the bug. 

Therefore forking inside a script intended to be run with the `runner` command on 7.1 crashes. (see #51298)

This PR is targeting `7-1-stable`

Edit: The regression test (with the fork) blew up the tests. I removed it, I'll revisit. 

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

skipkayhil commented:
I'm not sure this is the correct solution. The issue is the `ConnectionPool` gets `discard!`ed on `fork` here:

https://github.com/rails/rails/blob/e58d59ce489f7b7c48c0ace1b9d2ac03cb458258/activerecord/lib/active_record/connection_adapters/pool_config.rb#L82

and based on this comment

https://github.com/rails/rails/blob/e58d59ce489f7b7c48c0ace1b9d2ac03cb458258/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L284-L286

you really shouldn't be calling any methods on the `ConnectionPool` after calling `discard!`. So I think a proper fix would be checking for `discarded?` before whatever (the `Executor`?) is trying to interact with a discarded pool.

JoeDupuis committed:
Fix a crash when an  Executor wrapped fork exit.

Fix #51298

forking inside an Execution Wrapper crashes when running the completion callbacks.

Rails 7.0 was not Execution wrapping the `runner` command.
Rails 7.2 changed the definition of `active_connection?`, the new definition doesn't contain the bug.

Therefore forking inside a script intended to be run with the `runner` command on 7.1 crashes. (see #51298)

JoeDupuis commented:
@skipkayhil I was trying to catch all uses of `active_connection?` in case there was more Executor hooks after this one, but you are right, as you pointed out, the `QueryCache` should not rely on the discarded pool not blowing up as this is explicitly undefined.

I added a check for the status of the pool inside of `QueryCache` instead. What do you think?

JoeDupuis commented:
The failing test seems safe to ignore (unrelated). It's failing on the other `7-1-stable` builds:  https://buildkite.com/rails/rails/builds?branch=7-1-stable
